### DISCUSSION TITLE
Renovate stopped updating Java dependencies when using underscore in Gradle TOML version.ref

### DISCUSSION THREAD

**Discussion by JoergHeinicke5005** [renovatebot/renovate#22365] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Always latest, currently 35.98.4

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

Renovate should consider Java dependencies, specified in Gradle's TOML file using version.ref and an underscore in the version reference again.

### What you tried so far.

The trigger for finding and analyzing the issue was an expected but missing Spring Boot upgrade (to 2.7.12) in some of our repositories. For narrowing it down, I used a repository with AWS SDK dependency though as they are releasing almost daily.

Our libs.versions.toml is as follows:
```
[versions]
amazon_java_sdk          = "2.20.59"
nimbusds                 = "9.31"
slf4j_api                = "1.7.36"
spotbugs                 = "4.7.3"

[libraries]
amazonaws_java_ecs               = { module = "software.amazon.awssdk:ecs", version.ref = "amazon_java_sdk" }
amazonaws_java_secretsmanager    = { module = "software.amazon.awssdk:secretsmanager", version.ref = "amazon_java_sdk" }
nimbusds                         = { module = "com.nimbusds:nimbus-jose-jwt", version.ref = "nimbusds" }
slf4j_api                        = { module = "org.slf4j:slf4j-api", version.ref = "slf4j_api" }
spotbugs_annotations             = { module = "com.github.spotbugs:spotbugs-annotations", version.ref = "spotbugs" }
```

When checking, why expected changes are not picked up by Renovate, I was already missing the AWS SDK dependencies n the Renovate dashboard... When running Renovate in debug mode, I found a skipReason "no-version" for AWS SDK and SLF4j but a proper version for Nimbus DS and Spotbugs dependencies. From there it wasn't too hard to guess anymore: The underscore in the version names/ references are putting Renovate off.

After changing the version reference from `amazon_java_sdk` to `amazonjavasdk`, Renovate did identify the version correctly again and suggested an upgrade to AWS SDK 2.20.70 as expected.

We have Renovate running automatically multiple times a day from Monday to Thursday. The last successful run, updating AWS SDK to 2.20.59, was on May 4th at 10 AM UTC+2. Since we had no run from May 5th until May 7th, we only know it stopped working on May 8th at the latest.

### Relevant debug logs

<details><summary>Logs</summary>

```
                {
                 "depName": "amazonaws_java_secretsmanager",
                 "skipReason": "no-version",
                 "managerData": {"packageFile": "gradle/libs.versions.toml"},
                 "datasource": "maven",
                 "registryUrls": ["https://repo.maven.apache.org/maven2"],
                 "depType": "dependencies",
                 "updates": [],
                 "packageName": "amazonaws_java_secretsmanager"
               },
               {
                 "depName": "com.nimbusds:nimbus-jose-jwt",
                 "groupName": "nimbusds",
                 "currentValue": "9.31",
                 "managerData": {
                   "fileReplacePosition": 182,
                   "packageFile": "gradle/libs.versions.toml"
                 },
                 "fileReplacePosition": 182,
                 "datasource": "maven",
                 "registryUrls": ["https://repo.maven.apache.org/maven2"],
                 "depType": "dependencies",
                 "updates": [],
                 "packageName": "com.nimbusds:nimbus-jose-jwt",
                 "warnings": [],
                 "versioning": "gradle",
                 "sourceUrl": "https://bitbucket.org/connect2id/nimbus-jose-jwt",
                 "registryUrl": "https://repo.maven.apache.org/maven2",
                 "currentVersion": "9.31",
                 "fixedVersion": "9.31"
               },
```

</details>

**Comment by Churro** [comment]
This is most likely a regression due to #21820, which added "normalization" by filtering underscores. I'll try to see what gradle filters in its code but we may want to adjust this to just converting dots to dashes but not touching underscores.

**Reply by Churro** [reply, marked as answer]
PR #22470 intends to correct this.

The original idea of not touching underscores was too short-sighted. E.g., it is possible to declare `version.ref = "amazon_java_sdk"` but define the actual alias like this: `amazon-java-sdk = "2.20.59"`. The PR mimics what gradle does by ensuring that all lookups are done on the same normalized representation.

**Reply by JoergHeinicke5005** [reply]
Thanks a lot, @Churro. I can see in our repositories that Renovate analyses the dependencies correctly again.
